### PR TITLE
964-CreateHapiPluginToValidateErrorCodesOnPreHandler

### PR DESCRIPTION
**Pull request by gibaros**
Create a hapijs plugin to validate error codes onPreHandler.

NOTE: Currently only used at the ml-api-adapter at PUT /transfers/{id}/error route

### CODE REVIEW

File: src/handler.js [hunk 1]
```diff
@@ -99,3 +100,34 @@ exports.onPreResponse = function (request, reply) {
...
+  const response = request.response
+  let incomingErrorCode
+  if (response instanceof ErrorFactory.FSPIOPError || response.isBoom) {
+    try {
+      incomingErrorCode = response.toApiErrorObject.ErrorInformation.errorCode
+      if (!ErrorFactory.validateFSPIOPErrorCode(incomingErrorCode)) {
+        // TODO: validateFSPIOPErrorCode throws an exception if code is not valid at which point
+        //  we nee to call response.takeover()
+        // response.takeover(ErrorFactory.createFSPIOPError(Errors.VALIDATION_ERROR, `The incoming error code: ${incomingErrorCode} is not a valid mojaloop specification error code`))
+        return reply.response(ErrorFactory.createFSPIOPError(Errors.VALIDATION_ERROR, `The incoming error code: ${incomingErrorCode} is not a valid mojaloop specification error code`)).code(400)
```
Comment (mdebarros) [inline src/handler.js]: Two things here: 
1. I believe you need to add the `.takeover()` at the end of the calling stack
2. You will need to ensure that you `toApiErrorObject` the error is also called on the FSPIOPError assuming that the prehandler is skipped (which would have done this for you). Please verify if this is required.

File: src/handler.js [hunk 2]
```diff
@@ -99,3 +100,34 @@ exports.onPreResponse = function (request, reply) {
...
+      }
+    } catch (err) {
+      // TODO: validateFSPIOPErrorCode throws an exception if code is not valid at which point
+      //  we nee to call response.takeover()
+      // response.takeover(ErrorFactory.createFSPIOPError(Errors.VALIDATION_ERROR, `The incoming error code: ${incomingErrorCode} is not a valid mojaloop specification error code`))
+      return reply.response(ErrorFactory.createFSPIOPError(Errors.VALIDATION_ERROR, `The incoming error code: ${incomingErrorCode} is not a valid mojaloop specification error code`)).code(400)
```
Comment (mdebarros) [inline src/handler.js]: Point 2 above applies here as well.

File: src/handler.js [hunk 3]
```diff
@@ -99,3 +100,34 @@ exports.onPreResponse = function (request, reply) {
...
+ * @param request the http request
+ * @param reply
+ * @returns {boolean|reply.continue|continue|((key?: IDBValidKey) => void)}
+ */
+exports.onPreHandler = function (request, reply) {
+  const response = request.response
+  let incomingErrorCode
+  if (response instanceof ErrorFactory.FSPIOPError || response.isBoom) {
+    try {
+      incomingErrorCode = response.toApiErrorObject.ErrorInformation.errorCode
```
[the remark this answers is not in the document]
Comment (gibaros) [reply src/handler.js]: Addressed with last commit

File: src/handler.js (hunk 1 above)
Comment (gibaros) [reply src/handler.js]: Addressed with last commit

File: src/handler.js (hunk 2 above)
Comment (gibaros) [reply src/handler.js]: Addressed with last commit

**Pull request merged**